blockchain/reactor_test.go

```diff
@@ -0,0 +1,204 @@
+	state = sm.LoadState(stateDB)
+	state.SetLogger(stateLogger)
+
+	// Transaction indexing
+	var txIndexer txindex.TxIndexer
+	switch config.TxIndex {
+	case "kv":
+		store := db.NewDB("tx_index", config.DBBackend, config.DBDir())
+		txIndexer = kv.NewTxIndex(store)
+	default:
+		txIndexer = &null.TxIndex{}
+	}
+	state.TxIndexer = txIndexer
+
+	// Make event switch
+	eventSwitch := types.NewEventSwitch()
+	eventSwitch.SetLogger(logger.With("module", "types"))
+	_, err := eventSwitch.Start()
+	if err != nil {
+		cmn.Exit(cmn.Fmt("Failed to start switch: %v", err))
+	}
+
+	// Decide whether to fast-sync or not
+	// We don't fast-sync when the only validator is us.
+	fastSync := config.FastSync
+	if state.Validators.Size() == 1 {
+		addr, _ := state.Validators.GetByIndex(0)
+		if bytes.Equal(privValidator.Address, addr) {
+			fastSync = false
+		}
+	}
+
+	// Log whether this node is a validator or an observer
+	if state.Validators.HasAddress(privValidator.Address) {
+		consensusLogger.Info("This node is a validator")
+	} else {
+		consensusLogger.Info("This node is not a validator")
+	}
+
+	// Make BlockchainReactor
+	bcReactor := blockchain.NewBlockchainReactor(state.Copy(), proxyApp.Consensus(), blockStore, fastSync)
+	bcReactor.SetLogger(logger.With("module", "blockchain"))
+
+	return bcReactor, state
+}
+
+func newPeer() (*p2p.Peer, error) {
+	ed25519PubKey, _, err := ed25519.GenerateKey(rand.Reader)
+	if err != nil {
+		return nil, err
+	}
+	p := &p2p.Peer{
+		NodeInfo: &p2p.NodeInfo{
+			PubKey: crypto.PubKeyEd25519(*ed25519PubKey),
+		},
+		Key: "ABCDEFGHIJKLMNOPQRSTU",
+	}
+	return p, nil
+}
+
+func TestNoBlockMessageResponse(t *testing.T) {
+	logBuf := new(bytes.Buffer)
+	logger := log.NewTMLogger(logBuf)
+	bcr, state := newBlockchainReactor(logger)
+
+	peer, err := newPeer()
+	if err != nil {
+		t.Fatalf("malicious peer: %v", err)
+	}
+	bcr.AddPeer(peer)
+
+	chID := byte(0x2)
+	// For starters add a block of a known height
+	blockNumbers := []int{10, 20}
+
+	// Now let's make the blocks
+	for _, blockNumber := range blockNumbers {
+		block := makeBlock(blockNumber, state)
+
+		// Serialize that block now
+		buf := new(bytes.Buffer)
+		var n int
+		var err error
+		wire.WriteBinary(&struct{ Block *types.Block }{block}, buf, &n, &err)
+		// msgResponseBytes
+		msgResponseBytes := append([]byte{0x11}, buf.Bytes()...)
+		bcr.Receive(chID, peer, msgResponseBytes)
+	}
+
+	tests := [...]struct {
+		heightBytes []byte
+		nonExistent bool
+	}{
+		0: {
+			heightBytes: []byte{0x00, 0x28}, // height of 40
+			nonExistent: true,
+		},
+		1: {
+			heightBytes: []byte{0x00, byte(blockNumbers[0])}, // height of 10,
+			nonExistent: false,                               // We should have this
+		},
+	}
+
+	nonExistentSubstr := []byte("non-existent block")
+	for i, tt := range tests {
+		msgBytes := append([]byte{0x10, 0x02}, tt.heightBytes...)
+		bcr.Receive(chID, peer, msgBytes)
+		gotErr := bytes.Contains(logBuf.Bytes(), nonExistentSubstr)
```

Comment:
Oh boy. This is unfortunate. This is why I opened https://github.com/tendermint/tendermint/issues/631.